core/controllers/profile.py

```diff
@@ -134,71 +149,187 @@ def post(self, _):
+            'profile_picture_data_url': user_settings.profile_picture_data_url
+        })
+        self.render_json(self.values)
+
+    @acl_decorators.can_manage_own_account
+    def put(self):
+        """Handles PUT requests."""
+        update_type = self.normalized_payload.get('update_type')
+        data = self.normalized_payload.get('data')
+        if update_type == 'profile_picture_data_url':
+            user_services.update_profile_picture_data_url(self.user_id, data)
+
+
+class BulkEmailMessageHandler(base.BaseHandler):
+    """Provides bulk email signup messages."""
+
+    URL_PATH_ARGS_SCHEMAS = {}
+    HANDLER_ARGS_SCHEMAS = {
+        'GET': {},
+        'PUT': {
+            'update_type': {
+                'schema': {
+                    'type': 'basestring',
+                }
+            },
+            'data': {
+                'schema': {
+                    'type': 'dict',
+                },
+                'default_value': False
+            }
+        }
+    }
+
+    GET_HANDLER_ERROR_RETURN_TYPE = feconf.HANDLER_TYPE_JSON
+
+    @acl_decorators.can_manage_own_account
+    def get(self):
+        """Handles GET requests."""
+        user_email_preferences = user_services.get_email_preferences(
+            self.user_id)
+
+        self.values.update({
             'can_receive_email_updates': (
                 user_email_preferences.can_receive_email_updates),
             'can_receive_editor_role_email': (
                 user_email_preferences.can_receive_editor_role_email),
             'can_receive_feedback_message_email': (
                 user_email_preferences.can_receive_feedback_message_email),
             'can_receive_subscription_email': (
-                user_email_preferences.can_receive_subscription_email),
-            'subscription_list': subscription_list
+                user_email_preferences.can_receive_subscription_email)
         })
         self.render_json(self.values)
 
     @acl_decorators.can_manage_own_account
     def put(self):
         """Handles PUT requests."""
-        update_type = self.payload.get('update_type')
-        data = self.payload.get('data')
+        update_type = self.normalized_payload.get('update_type')
+        data = self.normalized_payload.get('data')
         bulk_email_signup_message_should_be_shown = False
+        if update_type == 'email_preferences':
+            bulk_email_signup_message_should_be_shown = (
+                user_services.update_email_preferences(
+                    self.user_id, data['can_receive_email_updates'],
+                    data['can_receive_editor_role_email'],
+                    data['can_receive_feedback_message_email'],
+                    data['can_receive_subscription_email']))
+        self.render_json({
+            'bulk_email_signup_message_should_be_shown': (
+                bulk_email_signup_message_should_be_shown)
+        })
+
+
+class PreferencesHandler(base.BaseHandler):
```

Comment:
For this handler I think we decided to send all the data all at once.

Reply:
As I am not working on this handler I removed them.